src/quotamodel.c

```diff
@@ -900,6 +897,75 @@ merge_uncommitted_table_to_oidlist(List *oidlist)
 	return oidlist;
 }
 
+static TableSizeEntry *
+get_tsentry(Oid oid, int16 segid)
+{
+	bool              found;
+	TableSizeEntryKey key     = {.reloid = oid, .id = TableSizeEntryId(segid)};
+	HASHACTION        action  = check_hash_fullness(table_size_map, MAX_NUM_TABLE_SIZE_ENTRIES, table_size_map_warning,
+	                                                table_size_map_last_overflow_report);
+	TableSizeEntry   *tsentry = hash_search(table_size_map, &key, action, &found);
+
+	if (!found && tsentry != NULL)
+	{
+		memset(tsentry->totalsize, 0, sizeof(tsentry->totalsize));
+		tsentry->owneroid      = InvalidOid;
+		tsentry->namespaceoid  = InvalidOid;
+		tsentry->tablespaceoid = InvalidOid;
+		tsentry->flag          = 0;
+
+		int seg_st = TableSizeEntrySegidStart(tsentry);
+		int seg_ed = TableSizeEntrySegidEnd(tsentry);
+		for (int j = seg_st; j < seg_ed; j++) TableSizeEntrySetFlushFlag(tsentry, j);
+	}
+
+	/* mark tsentry is_exist */
+	if (tsentry) set_table_size_entry_flag(tsentry, TABLE_EXIST);
+
+	return tsentry;
+}
+
+void
+update_active_table_size(Oid oid, int64 size, int16 segid)
+{
+	TableSizeEntry *tsentry = get_tsentry(oid, segid);
+
+	if (tsentry == NULL)
+	{
+		/* Too many tables have been added to the table_size_map, to avoid diskquota using
+		   too much share memory, just return. The diskquota won't work correctly
+		   anymore. */
+		return;
+	}
+
+	if (segid == -1)
+	{
+		/* pretend process as utility mode, and append the table size on master */
+		Gp_role = GP_ROLE_UTILITY;
```

Comment:
why do we need to pretend to be an utility?

Reply:
This has nothing to do with the current patch.

Reply:
Please explain.

Reply:
This was also the case before the patch.

Reply:
- "Why do we need..." ?
- "it was before the patch!"
It got into the diff and raised questions. May be get an answer or restructure the change so it won't get into diff then?

Reply:
Well, I was counting on the reviewer to still compare the new functions with the pieces of code that were cut.

Reply:
It seems that this was necessary to avoid dispatching the `pg_total_relation_size` function, which was previously called instead of the `calculate_table_size` function at this location.

Reply:
According to this, it is not needed anymore, right?

Reply:
Maybe, but it shouldn't be solved in the current patch! You can create a separate task for this.